core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala

```diff
@@ -0,0 +1,338 @@
+      }
+    }
+
+    def transform(plan: LogicalPlan): LogicalPlan =
+      transformSubqueries(plan) transform {
+        case scan @ DeltaTableScan(projection, filters, fileIndex, limit, delta) =>
+          val scanGenerator = getDeltaScanGenerator(fileIndex)
+          val (scannedSnapshot, preparedScan) = deltaScans.getOrElseUpdate(scan.canonicalized,
+              filesForScan(scanGenerator, limit, projection, filters, delta))
+          val preparedIndex = getPreparedIndex(preparedScan, fileIndex)
+          optimizeGeneratedColumns(
+            scannedSnapshot, scan, preparedIndex, filters, limit, delta)
+      }
+
+    transform(plan)
+  }
+
+  protected def optimizeGeneratedColumns(
+      scannedSnapshot: Snapshot,
+      scan: LogicalPlan,
+      preparedIndex: PreparedDeltaFileIndex,
+      filters: Seq[Expression],
+      limit: Option[Int],
+      delta: LogicalRelation): LogicalPlan = {
+    // TODO: future generated columns optimization
+    DeltaTableUtils.replaceFileIndex(scan, preparedIndex)
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (spark.sessionState.conf.getConf(DeltaSQLConf.DELTA_STATS_SKIPPING)) {
+      var metadataOptimizedPlan = plan
+
+      // Should not be applied to subqueries to avoid duplicate delta jobs.
+      val isSubquery = plan.isInstanceOf[Subquery] || plan.isInstanceOf[SupportsSubquery]
+      // Should not be applied to DataSourceV2 write plans, because they'll be planned later
+      // through a V1 fallback and only that later planning takes place within the transaction.
+      val isDataSourceV2 = plan.isInstanceOf[V2WriteCommand]
+      if (isSubquery || isDataSourceV2) {
+        return metadataOptimizedPlan
+      }
+
+      prepareDeltaScan(metadataOptimizedPlan)
+    } else {
+      // If this query is running inside an active transaction and is touching the same table
+      // as the transaction, then mark that the entire table as tainted to be safe.
+      OptimisticTransaction.getActive.foreach { txn =>
+        val logsInPlan = plan.collect { case DeltaTable(fileIndex) => fileIndex.deltaLog }
+        if (logsInPlan.exists(_.isSameLogAs(txn.deltaLog))) {
+          txn.readWholeTable()
+        }
+      }
+
+      // Just return the plan if statistics based skipping is off.
+      // It will fall back to just partition pruning at planning time.
+      plan
+    }
+  }
+
+  object DeltaTableScan {
+
+    /**
+     * The components of DeltaTableScanType are:
+     * - an `AttributeSet` of the project collected by `PhysicalOperation`
+     * - filter expressions collected by `PhysicalOperation`
+     * - the `TahoeLogFileIndex` of the matched DeltaTable`
+     * - integer value of limit expression, if any
+     * - matched `DeltaTable`
+     */
+    private type DeltaTableScanType =
+      (Seq[Attribute], Seq[Expression], TahoeLogFileIndex, Option[Int], LogicalRelation)
+
+    def unapply(plan: LogicalPlan): Option[DeltaTableScanType] = {
```

Comment:
interesting way to use scala pattern matching.... i like it. but document this better on what you are trying to match.